tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/json/JsonToTlaViaBuilder.scala

```diff
@@ -0,0 +1,274 @@
+
+  private def asFParam(json: T): OperParam = {
+    val kindField = getOrThrow(json, TlaToJson.kindFieldName)
+    val kind = scalaFactory.asStr(kindField)
+
+    kind match {
+      case "OperParam" =>
+        val nameField = getOrThrow(json, "name")
+        val name = scalaFactory.asStr(nameField)
+        val arityField = getOrThrow(json, "arity")
+        val arity = scalaFactory.asInt(arityField)
+        OperParam(name, arity)
+      case _ => throw new JsonDeserializationError(s"$kind is not a valid OperParam kind")
+    }
+  }
+
+  private def asTlaValEx(json: T): TBuilderInstruction = {
+    val kindField = getOrThrow(json, TlaToJson.kindFieldName)
+    val kind = scalaFactory.asStr(kindField)
+
+    kind match {
+      case "TlaStr" =>
+        val valField = getOrThrow(json, "value")
+        val valStr = scalaFactory.asStr(valField)
+        tla.str(valStr)
+      case "TlaDecimal" =>
+        throw new JsonDeserializationError("TlaDecimal is not supported.")
+      case "TlaInt" =>
+        val valField = getOrThrow(json, "value")
+        val bi = valField
+          .getFieldOpt("bigInt")
+          .map { biField =>
+            BigInt(scalaFactory.asStr(biField))
+          }
+          .getOrElse {
+            BigInt(scalaFactory.asInt(valField))
+          }
+        tla.int(bi)
+      case "TlaBool" =>
+        val valField = getOrThrow(json, "value")
+        val valBool = scalaFactory.asBool(valField)
+        tla.bool(valBool)
+      case "TlaBoolSet" => tla.booleanSet()
+      case "TlaIntSet"  => tla.intSet()
+      case "TlaStrSet"  => tla.stringSet()
+      case "TlaNatSet"  => tla.natSet()
+      case _            => throw new JsonDeserializationError(s"$kind is not a valid TlaValue kind")
+    }
+  }
+
+  private def getSourceLocationOpt(json: T): Option[SourceLocation] = for {
+    sourceObj <- json.getFieldOpt(TlaToJson.sourceFieldName)
+    fileName <- sourceObj.getFieldOpt("filename").map(scalaFactory.asStr)
+    fromObj <- sourceObj.getFieldOpt("from")
+    toObj <- sourceObj.getFieldOpt("to")
+    fromLine <- fromObj.getFieldOpt("line").map(scalaFactory.asInt)
+    fromCol <- fromObj.getFieldOpt("column").map(scalaFactory.asInt)
+    toLine <- toObj.getFieldOpt("line").map(scalaFactory.asInt)
+    toCol <- toObj.getFieldOpt("column").map(scalaFactory.asInt)
+  } yield SourceLocation(
+      fileName,
+      SourceRegion(
+          SourcePosition(fromLine, fromCol),
+          SourcePosition(toLine, toCol),
+      ),
+  )
+
+  private def setLoc(identifiable: Identifiable, locationOpt: Option[SourceLocation]): Unit =
+    sourceStoreOpt.foreach { store =>
+      locationOpt.foreach { loc =>
+        store.add(identifiable.ID, loc)
+      }
+    }
+
+  override def asTlaModule(moduleJson: T): TlaModule = rethrowAsJsonDeserializationException {
+    val kindField = getOrThrow(moduleJson, TlaToJson.kindFieldName)
```

Comment:
This `rethrow` has much larger scope than necessary

Reply:
I agree, but try-catch doesn't work well with `State[_]`. Essentially, the exception is only thrown when `build` is called, but that's implicit. A "tight" solution would involve making the build calls explicit, and doing try-catch there, which I feel exposes the guts of the builder a bit too much.

Reply:
I don't like it, but then let's keep it.
Can you add a comment why we're doing it this way?

Reply:
I could also try something else: instead of importing the implicit build, I could redefine my own, that does try-catch at build-time.

Reply:
Either sounds fine